genet/core.py

```diff
@@ -2263,3 +2264,48 @@ def get_link_slope_dictionary(self, elevation_dict):
             slope_dict[link_id] = {'slope': link_slope}
 
         return slope_dict
+
+    def split_link_at_point(self, link_id, point_lat, point_lon, new_node_id):
+        """
+        Takes a link and point coordinates, and splits the link at the point to create 2 new links;
+        the old link is then deleted.
+        :param link_id: ID of the link to split
+        :param point_lat: latitude of the point to split at
+        :param point_lon: longitude of the point to split at
+        :param new_node_id: ID to give to the new node
+        :return: self
+        """
+
+        # check if point is on the link LineString
+        point = Point(point_lon, point_lat)
+        from_node = self.link(link_id)['from']
+        to_node = self.link(link_id)['to']
+        from_node = self.node(from_node)
+        to_node = self.node(to_node)
+        line = LineString([(float(from_node['lon']), float(from_node['lat'])),
+                           (float(to_node['lon']), float(to_node['lat']))])
+
+        # if not, find nearest point on the link line
+        # (not using 'contains' method due to too high accuracy required to evaluate to True)
+        if line.distance(point) > 1e-8:
+            point = line.interpolate(line.project(point))
+
+        self.add_node(new_node_id, {'id': new_node_id, 'lon': point.x, 'lat': point.y})
+
+        # create 2 new links: from_node -> new_node ; new_node -> to_node
+        new_link_1 = link_id + '_1'
+        new_link_2 = link_id + '_2'
```

Comment:
not a huge fan of these, especially if the same link gets divided multiple times, it'll end up looking really crazy, use https://github.com/arup-group/genet/blob/4fcc3d36a49f8b042e7987082802ee61f3c2fc34/genet/core.py#L1945 to generate unique, unsused IDs